jwplatform/upload.py

```diff
@@ -0,0 +1,137 @@
+
+        # Upload the parts
+        for part_number in range(1, part_count + 1):
+            bytes_chunk = self.file.read(self.target_part_size)
+            if part_number < part_count and len(bytes_chunk) != self.target_part_size:
+                raise IOError("Failed to read enough bytes")
+            retry_count = 0
+            while retry_count < self.upload_retry_count:
+                try:
+                    self._upload_part(bytes_chunk, part_number, upload_links)
+                    break
+                except (DataIntegrityError, HTTPError):
+                    self.logger.warning(f"Encountered error upload part {part_number} of {part_count} for file {filename}."
+                                        f" Retrying.")
+                    retry_count = retry_count + 1
+
+            if retry_count >= self.upload_retry_count:
+                raise MaxRetriesExceededError(f"Max retries ({self.upload_retry_count}) exceeded while uploading part"
+                                              f" {part_number} of {part_count} for file {filename}.")
+
+        # Mark upload as complete
+        self._mark_upload_completion()
+
+    def _upload_part(self, bytes_chunk, part_number, upload_links):
+        # Add a S3 server-side checksum validation too if possible.
+        computed_hash = self._compute_part_hash(bytes_chunk)
+
+        # Check if the file has already been uploaded and the hash matches. Return immediately without doing anything
+        # if the hash matches.
+        upload_hash = upload_links[part_number - 1]["etag"] if "etag" in upload_links[part_number - 1] else None
+        if upload_hash:
+            if repr(upload_hash) == repr(f"\"{computed_hash}\""):  # The returned hash is surrounded by '"' character
+                return
+
+        upload_link = upload_links[part_number - 1]["upload_link"] if "upload_link" in upload_links[part_number - 1] \
+            else None
+        if not upload_link:
+            raise Exception(f"Invalid upload link for part {part_number}.")
+
+        resp = requests.put(upload_links[part_number - 1]["upload_link"], data=bytes_chunk)
+        resp.raise_for_status()
+
+        returned_hash = resp.headers['ETag']
+        if repr(returned_hash) != repr(f"\"{computed_hash}\""):  # The returned hash is surrounded by '"' character
+            raise DataIntegrityError("The hash of the uploaded file does not match with the hash on the server.")
+        self.logger.info(f"Successfully uploaded part {part_number} for upload id {self.upload_id}")
+
+    def _get_pre_signed_part_links(self, part_count) -> {}:
+        query_params = {'page_length': part_count}
+
+        resp = self.client.list(resource_name='uploads', resource_id=self.upload_id, subresource_name='parts',
+                                query_params=query_params)
+
+        body = resp.json_body
+        # Process the results if there are multiple pages
```

Comment:
This could be memory-intensive and error-prone for files with many parts. It might be better to upload a full page of parts before moving onto the next.

Reply:
The max number of parts is limited to 10,000, and we could always retry in the case of errors, so I'm not sure these are huge concerns. I think the major reason to do the full upload flow in batches is mostly so that upload progress is made sooner, rather than having to wait to retrieve all the parts upfront.

Reply:
The logic has been inverted to incorporate the recommendation.